compiler/lib/global_deadcode.ml

```diff
@@ -0,0 +1,435 @@
+  in
+  Addr.Map.iter (fun _ block -> live_block block) prog.blocks;
+  live_vars
+
+(* Returns the set of variables given the adjacency list of variable dependencies. *)
+let variables deps =
+  let vars = Var.ISet.empty () in
+  Array.iteri ~f:(fun i _ -> Var.ISet.add vars (Var.of_idx i)) deps;
+  vars
+
+(** Propagate liveness of the usages of a variable [x] to [x]. The liveness of [x] is
+    defined by joining its current liveness and the contribution of each vairable [y]
+    that uses [x]. *)
+let propagate uses defs live_vars live_table x =
+  let idx = Var.idx x in
+  (* Variable [y] uses [x] either in its definition ([Compute]) or as a closure/block parameter
+      ([Propagate]). In the latter case, the contribution is simply the liveness of [y]. In the former,
+       the contribution depends on the liveness of [y] and its definition. *)
+  let contribution y usage_kind =
+    match usage_kind with
+    (* If x is used to compute y, we consider the liveness of y *)
+    | Compute -> (
+        match Var.Tbl.get live_table y with
+        (* If y is dead, then x is dead. *)
+        | Dead -> Dead
+        (* If y is a live block, then x is the join of liveness fields that are x *)
+        | Live fields -> (
+            match defs.(Var.idx y) with
+            | Expr (Block (_, vars, _)) ->
+                let found = ref false in
+                Array.iteri
+                  ~f:(fun i v ->
+                    if Var.equal v x && IntSet.mem i fields then found := true)
+                  vars;
+                if !found then Top else Dead
+            | Expr (Field (_, i)) -> Live (IntSet.singleton i)
+            | _ -> Top)
+        (* If y is top and y is a field access, x depends only on that field *)
+        | Top -> (
+            match defs.(Var.idx y) with
+            | Expr (Field (_, i)) -> Live (IntSet.singleton i)
+            | _ -> Top))
+    (* If x is used as an argument for parameter y, then contribution is liveness of y *)
+    | Propagate -> Var.Tbl.get live_table y
+  in
+  Var.Map.fold
+    (fun y usage_kind live -> Domain.join (contribution y usage_kind) live)
+    uses.(idx)
+    (Domain.join live_vars.(idx) (Var.Tbl.get live_table x))
+
+let solver vars uses defs live_vars =
+  let g =
+    { G.domain = vars
+    ; G.iter_children = (fun f x -> Var.Map.iter (fun y _ -> f y) uses.(Var.idx x))
+    }
+  in
+  Solver.f () (G.invert () g) (propagate uses defs live_vars)
+
+(** Replace each instance of a dead variable with a sentinal value. 
+  Blocks that end in dead variables are compacted to the first live entry. 
+  Dead variables are replaced when
+    + They appear in a dead field of a block; or
+    + They are returned; or
+    + They are applied to a function. 
+ *)
+let zero prog sentinal live_table =
+  let compact_vars vars =
+    let i = ref (Array.length vars - 1) in
+    while !i >= 0 && Var.equal vars.(!i) sentinal do
+      i := !i - 1
+    done;
+    Array.sub vars ~pos:0 ~len:(!i + 1)
```

Comment:
We can avoid the array allocation if `!i + 1` = `length`

Reply:
[0214876](https://github.com/ocsigen/js_of_ocaml/pull/1503/commits/02148767e80b11152e311c0f437f8d5ba59f386f)